File: packages/ai-setup-go/internal/scaffold/specs.go
package scaffold

import (
	"io/fs"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/ricardoborges-teachable/ai-setup/internal/files"
	"github.com/ricardoborges-teachable/ai-setup/internal/types"
)

var specifyTemplates = []string{
	"spec-template.md",
	"plan-template.md",
	"tasks-template.md",
	"checklist-template.md",
	"task-harness-template.md",
}

// HasSpecKitStructure returns true if a .specify/ directory already exists.
func HasSpecKitStructure(targetDir string) bool {
	return files.IsDirectory(filepath.Join(targetDir, ".specify"))
}

// DetectExistingSpecs scans specs/ for ###-slug directories and returns
// the highest number found (0 if none).
func DetectExistingSpecs(targetDir string) (hasSpecs bool, highest int) {
	specsDir := filepath.Join(targetDir, "specs")
	if !files.IsDirectory(specsDir) {
		return false, 0
	}

	entries, err := fs.ReadDir(files.OSDirFS(specsDir), ".")
	if err != nil {
		return false, 0
	}

	highest = 0
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		// Match ###- pattern (e.g., "001-feature-name")
		name := entry.Name()
		if len(name) >= 4 && name[3] == '-' {
			if num, err := strconv.Atoi(name[:3]); err == nil && num > highest {
				highest = num
			}
		}
	}

	return highest > 0, highest
}

// ScaffoldSpecs creates the speckit-compatible .specify/ and specs/ directory structure.
//
// Behavior:
// - If .specify/ already exists: skip .specify/ scaffolding entirely (respect existing)
// - If specs/###-slug/ directories exist: skip spec directory creation
// - Greenfield: create specs/.gitkeep + full .specify/ with templates, memory, scripts
func ScaffoldSpecs(targetDir string, setupScope types.SetupScope, libFS fs.FS, specsDirs []string, fileRecords *[]types.TrackedFile, strategy types.ConflictStrategy, perFileOverrides map[string]types.ConflictStrategy) error {
	_ = libFS
	_ = fileRecords
	_ = strategy
	_ = perFileOverrides

	hasSpecify := HasSpecKitStructure(targetDir)
	_, hasSpecs := DetectExistingSpecs(targetDir)

	// 1. Create .specify/ directory (speckit core) — only if not already present.
	if !hasSpecify {
		specifyDir := filepath.Join(targetDir, ".specify")

		// 1a. .specify/templates/
		templatesDir := filepath.Join(specifyDir, "templates")
		_ = files.EnsureDir(templatesDir)
		// Templates are copied by the TS runtime via the library; Go creates
		// the directory structure. The actual file copy happens in the
		// compile/adapter layer which reads library/templates/.
		for _, tpl := range specifyTemplates {
			_ = tpl // placeholder — actual copy handled by adapter layer
		}

		// 1b. .specify/memory/
		memoryDir := filepath.Join(specifyDir, "memory")
		_ = files.EnsureDir(memoryDir)

		// Workspace: create repos/ ledger directory.
		if setupScope == types.SetupScopeWorkspace {
			_ = files.EnsureDir(filepath.Join(memoryDir, "repos"))
		}

		// 1c. .specify/scripts/bash/
		_ = files.EnsureDir(filepath.Join(specifyDir, "scripts", "bash"))
	}

	// 2. Create specs/ directory.
	specsDir := filepath.Join(targetDir, "specs")
	_ = files.EnsureDir(specsDir)

	// Greenfield: create .gitkeep.
	if !hasSpecs && len(specsDirs) == 0 {
		gitkeepPath := filepath.Join(specsDir, ".gitkeep")
		if err := files.WriteFile(gitkeepPath, []byte("# Specs\n\nFeature specifications are created by the `/speckit.specify` command.\n"), 0o644); err == nil {
			hash, _ := files.FileHash(gitkeepPath)
			if fileRecords != nil {
				relPath, _ := filepath.Rel(targetDir, gitkeepPath)
				*fileRecords = append(*fileRecords, types.TrackedFile{
					Path:   filepath.ToSlash(relPath),
					Hash:   hash,
					Source: "speckit:specs-root",
					Owner:  types.FileOwnerLibrary,
				})
			}
		}
	}

	// 3. Legacy: create selected specs directories.
	for _, dir := range specsDirs {
		dirPath := filepath.Join(specsDir, dir)
		if err := files.EnsureDir(dirPath); err != nil {
			return err
		}

		// Special case: memory also needs handoffs subdirectory.
		if dir == "memory" {
			if setupScope == types.SetupScopeWorkspace {
				_ = files.EnsureDir(filepath.Join(specsDir, "memory", "decisions"))
				_ = files.EnsureDir(filepath.Join(specsDir, "memory", "patterns"))
				_ = files.EnsureDir(filepath.Join(specsDir, "memory", "projects"))
			}
			_ = files.EnsureDir(filepath.Join(specsDir, "memory", "handoffs"))
		}
	}

	return nil
}

// OSDirFS adapter for reading directory entries from the OS filesystem.
// This avoids importing os directly in scaffold helpers.
var osDirFS = func(path string) fs.FS {
	dirFS := files.OSDirFS(path)
	return dirFS
}

File: packages/ai-setup-ts/src/scaffold/specs.ts
import * as fs from 'node:fs'
import path from 'node:path'
import type { ConflictStrategy, FileRecord, SetupScope } from '../types.js'
import { applyStrategy } from '../utils/conflict-strategy.js'
import { copyFile, ensureDir, fileExists, fileHash, isDirectory, writeFile } from '../utils/files.js'

export interface ScaffoldSpecsOptions {
  targetDir: string
  setupScope?: SetupScope
  libraryDir: string
  specsDirs: string[]
  specsAgents: string[]
  fileRecords: FileRecord[]
  strategy: ConflictStrategy
  perFileOverrides: Map<string, ConflictStrategy>
}

const SPECIFY_TEMPLATES = [
  'spec-template.md',
  'plan-template.md',
  'tasks-template.md',
  'checklist-template.md',
  'task-harness-template.md',
]

const SPECIFY_SCRIPTS = ['create-new-feature.sh']

const STARTER_STANDARD_FILES = [
  'agent-security.md',
  'context-loading.md',
  'error-handling.md',
  'orchestration-patterns.md',
  'test-patterns.md',
]

/**
 * Detects whether an existing spec-kit structure is present.
 * Returns true if .specify/ directory exists.
 */
export function hasSpecKitStructure(targetDir: string): boolean {
  return isDirectory(path.join(targetDir, '.specify'))
}

/**
 * Detects numbered spec directories (###-slug pattern).
 * Returns the highest existing number, or 0 if none.
 */
export function detectExistingSpecs(targetDir: string): { hasSpecs: boolean; highestNumber: number } {
  const specsDir = path.join(targetDir, 'specs')
  if (!isDirectory(specsDir)) return { hasSpecs: false, highestNumber: 0 }

  let highest = 0
  const entries = fileExists(specsDir) ? fs.readdirSync(specsDir, { withFileTypes: true }) : []
  for (const entry of entries) {
    if (!entry.isDirectory()) continue
    const match = entry.name.match(/^(\d{3})-/)
    if (match?.[1]) {
      const num = Number.parseInt(match[1], 10)
      if (num > highest) highest = num
    }
  }

  return { hasSpecs: highest > 0, highestNumber: highest }
}

/**
 * Creates the speckit-compatible .specify/ and specs/ directory structure.
 *
 * Behavior:
 * - If `.specify/` already exists: skip .specify/ scaffolding entirely (respect existing)
 * - If `specs/###-slug/` directories exist: skip spec directory creation
 * - Greenfield: create `specs/.gitkeep` + full `.specify/` with templates, memory, scripts
 * - Workspace mode also creates `.specify/memory/repos/` for ledger files
 * - Always copies template files from library to `.specify/templates/`
 */
export async function scaffoldSpecs(opts: ScaffoldSpecsOptions): Promise<void> {
  const { targetDir, setupScope, libraryDir, specsDirs, specsAgents, fileRecords, strategy, perFileOverrides } = opts

  const hasSpecify = hasSpecKitStructure(targetDir)
  const existing = detectExistingSpecs(targetDir)

  // 1. Create .specify/ directory (speckit core)
  if (!hasSpecify) {
    const specifyDir = path.join(targetDir, '.specify')

    // 1a. .specify/templates/ — copy template files from library
    const templatesDir = path.join(specifyDir, 'templates')
    ensureDir(templatesDir)
    for (const tplFile of SPECIFY_TEMPLATES) {
      const src = path.join(libraryDir, 'templates', tplFile)
      const dest = path.join(templatesDir, tplFile)
      await copyLibraryFile(src, dest, fileRecords, targetDir, strategy, perFileOverrides)
    }

    // 1b. .specify/memory/ — constitution is written by scaffoldConstitution
    //     (with placeholder substitution); just ensure the dir + workspace repos.
    const memoryDir = path.join(specifyDir, 'memory')
    ensureDir(memoryDir)

    // Workspace: create repos/ ledger directory
    if (setupScope === 'workspace') {
      ensureDir(path.join(memoryDir, 'repos'))
    }

    // 1c. .specify/scripts/
    if (SPECIFY_SCRIPTS.length > 0) {
      const scriptsDir = path.join(specifyDir, 'scripts', 'bash')
      ensureDir(scriptsDir)
    }
  }

  // 2. Create specs/ directory
  const specsDir = path.join(targetDir, 'specs')
  ensureDir(specsDir)

  // Greenfield: create .gitkeep to signal empty spec directory
  if (!existing.hasSpecs && specsDirs.length === 0) {
    const gitkeepPath = path.join(specsDir, '.gitkeep')
    const action = applyStrategy(gitkeepPath, strategy, perFileOverrides, targetDir)
    if (action !== 'skip') {
      writeFile(gitkeepPath, '# Specs\n\nFeature specifications are created by the `/speckit.specify` command.\n')
      fileRecords.push({
        path: path.relative(targetDir, gitkeepPath).replaceAll('\\', '/'),
        hash: fileHash(gitkeepPath),
        source: 'speckit:specs-root',
        owner: 'library',
      })
    }
  }

  // Seed starter standards file-by-file. Metadata files such as .gitkeep or
  // README.md must not suppress missing standards, and existing user-authored
  // files at the same paths are always preserved regardless of conflict strategy.
  copyStarterStandards(libraryDir, targetDir, fileRecords)

  // Legacy: create selected specs directories (only if existing spec-kit structure AND user selected them)
  if (specsDirs.length > 0) {
    for (const dir of specsDirs) {
      ensureDir(path.join(specsDir, dir))

      if (dir === 'memory') {
        if (setupScope === 'workspace') {
          ensureDir(path.join(specsDir, 'memory', 'decisions'))
          ensureDir(path.join(specsDir, 'memory', 'patterns'))
          ensureDir(path.join(specsDir, 'memory', 'projects'))
        }
        ensureDir(path.join(specsDir, 'memory', 'handoffs'))
      }
    }
  }

  // 3. Copy AGENTS.md files for selected specs directories (unchanged behavior)
  for (const dir of specsAgents) {
    const src = path.join(libraryDir, 'specs-agents', `${dir}.md`)
    const dest = path.join(specsDir, dir, 'AGENTS.md')

    await copyLibraryFile(src, dest, fileRecords, targetDir, strategy, perFileOverrides)
  }
}

function copyStarterStandards(libraryDir: string, targetDir: string, records: FileRecord[]): void {
  const starterDir = path.join(targetDir, 'specs', 'standards', 'starter')

  for (const standard of STARTER_STANDARD_FILES) {
    const sourcePath = path.join(libraryDir, 'standards', 'starter', standard)
    if (!fileExists(sourcePath)) continue

    const destPath = path.join(starterDir, standard)
    if (fileExists(destPath)) continue

    copyFile(sourcePath, destPath)
    records.push({
      path: path.relative(targetDir, destPath).replaceAll('\\', '/'),
      hash: fileHash(destPath),
      source: `standards:standards/starter/${standard}`,
      owner: 'library',
    })
  }
}

/**
 * Helper: Copy a file from library with conflict resolution and recording.
 */
async function copyLibraryFile(
  src: string,
  dest: string,
  records: FileRecord[],
  targetDir: string,
  strategy: ConflictStrategy,
  perFileOverrides: Map<string, ConflictStrategy>,
): Promise<void> {
  if (!fileExists(src)) return

  const relPath = path.relative(targetDir, dest)
  const action = applyStrategy(dest, strategy, perFileOverrides, targetDir)

  if (action === 'skip') {
    console.warn(`⚠️  Skipping existing file: ${relPath}`)
    return
  }

  copyFile(src, dest)
  records.push({
    path: relPath,
    hash: fileHash(dest),
    source: path.relative(targetDir, src),
    owner: 'library',
  })
}
